### ISSUE TITLE
Log plots support

### ISSUE THREAD

**Issue by florian6973** [predict-idlab/plotly-resampler#206]
Hi!

I am trying to use plotly-resampler for log plots, but since it is a linear downsampling, it does not work (bad resolution on the left on the graph). Would you know a workaround by any chance, or if it is a planned feature?

Thanks!

**Comment by jonasvdd** [comment]
@florian6973, If you use plotly-resampler log support branch see #207, would the code snippets added in the comments resolve your issue?

Main things I did:
- the pull-request fixes the zoom-resample functionality for log-axis plots
- I altered an aggregation algorithm to use log-based equidistant bins to select datapoints.

Kind regards, 
Jonas.

**Comment by florian6973** [comment]
Hi Jonas!
Wow thanks for working on this feature so quickly! 
I am sorry, I should have been more precise in my first message, but I was mainly interested in plotting the y-axis with a logarithmic scale, and it seems the code snippets work for x axis only (I think it can be really useful too in any case ;))
![image](https://user-images.githubusercontent.com/70778912/236686130-c3c9eb2e-fabd-4f17-9c5f-473f79e7a4b8.png)
Just out of curiosity, what does the 'NoGapHandler' class do please?
Thanks again :)

**Comment by jonasvdd** [comment]
Hi @florian6973, no problem at all! You had an interesting use-case, and this functionality was not yet supported by plotly-resampler, so was certainly worthwhile to investigate!

>  Just out of curiosity, what does the 'NoGapHandler' class do please?

If you read this part of the [documentation](https://predict-idlab.github.io/plotly-resampler/aggregation.html#module-plotly_resampler.aggregation.gap_handlers), you will see that the `NoGapHandler` does not introduce gaps (i.e. unconnected lines) in the plot. 
Side questions:
- **How are gaps determined**: Gaps are determined through assessing the differences between consecutive x-values of the aggregated series. If this difference is above a certain threshold, which is how the `MedDiffGapHandler` works, a gap will be inserted.
- **Why is this important for log-x aggregators**: Log-x aggregators such as `LogLTTB` will perform binning to ensure equidistant log-x bins. This will create large irregularities in the *non-log* x-space; making the default x-space based gap handlers introduce gaps where there shouldn't be any.  Setting the gap handler to `NoGapHandler` is an ad-hoc solution to alleviate this unwanted behavior. 

I can see that the second demo visualizes the data via a log x and log-y? So If you would use the `NoGapHandler` (and maybe increase the number of output samples (i.e. `n_shown_samples`), the visualization should be fine?

fyi: Could you send me some example data from your log-log FT plot? I think that would make a good use-case to add in the examples.

Kind regards,
Jonas

**Comment by jonasvdd** [comment]
Mhmm, I am not following you 100%; I just loaded the `dat_s.txt` but this does not give me the plot which you show above. 

Can you give me some example data and code where your log gap issue occurs, even when the nogaphandler is set?

**Comment by florian6973** [comment]
Yes you're right, sorry I mixed up the two examples. I get the graph below when I plot the FT of the signal from `dat_s.txt` (NoGapHandler is working well in this case):
_figure_:
![image](https://github.com/predict-idlab/plotly-resampler/assets/70778912/5d959296-91b6-472d-9089-0208ca401ccb)
_code_:
```python
import numpy as np
from plotly_resampler.aggregation import NoGapHandler
from plotly_resampler import FigureResampler
from plotly_resampler.aggregation.aggregation_interface import DataPointSelector
from typing import Union
import plotly.graph_objects as go

[... LogLLTB class ...]

file = "data.npy"
dat_s = np.load(file)

fft_coefs = np.abs(np.fft.rfft(dat_s))
fft_freq = np.fft.rfftfreq(len(dat_s), d=1 / 30000)

fr = FigureResampler(False)
fr.add_trace(
    go.Scattergl(mode="lines+markers"),
    hf_x=fft_freq,
    hf_y=fft_coefs,
    downsampler=LogLTTB(),
    gap_handler=NoGapHandler(),
    max_n_samples=10000,
)
fr.update_xaxes(type="log")
fr.update_yaxes(type="log")
fr.update_layout(template="plotly_white", title="log axis demo")
fr.show_dash(mode="inline")
```

However I still have an issue when I interactively zoom in on the figure: the curve disappear, as you can see:
![newplot (16)](https://github.com/predict-idlab/plotly-resampler/assets/70778912/450e48b2-be05-4aed-b43a-354569d3cb71)

Concerning the issue about NoGapHandler, it is with this example:
```python
[... imports and LogLTTB class ...]

n = 100_000
y = np.sin(np.arange(n) / 2_000) + np.random.randn(n) / 10

fr = FigureResampler(False)
fr.add_trace(
    go.Scattergl(mode="lines+markers", marker_color=np.abs(y) / np.max(np.abs(y))),
    hf_x=np.arange(n),
    hf_y=y,
    downsampler=LogLTTB(),
    gap_handler=NoGapHandler(),
    max_n_samples=10000,
)
fr.update_xaxes(type="log")
fr.update_yaxes(type="log")
fr.update_layout(template="plotly_white", title="log axis demo")

fr.show_dash(mode='inline')
```
This gives the figure of my previous message, and I am surprised there are gaps between points, even though it is not supposed to be so.

Thanks :)

**Comment by jonasvdd** [comment]
I think this issue can be closed if (i) the log plot PR gets merged and (ii) the above answer resolved your question!

**Comment by florian6973** [comment]
Thank you very much Jonas for all your explanations and help! 
Indeed I cloned the repo, switched to the log_support branch and tested, it works perfectly then! So I will wait for the merging of the PR and then its release to use this feature ;) 
Oh my bad, to be honest I did not look at the values and just applied a log axis 😅  By the way I always thought I would get at least a numpy warning if I take the log of a negative number... 
Thanks again!
